stl/inc/xmemory

```diff
@@ -677,35 +740,73 @@ struct _Default_allocator_traits { // traits for std::allocator
     template <class _Other>
     using rebind_traits = allocator_traits<allocator<_Other>>;
 
-    _NODISCARD static __declspec(allocator) pointer allocate(_Alloc&, _CRT_GUARDOVERFLOW const size_type _Count) {
-        return static_cast<pointer>(_Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
+        allocate(_Alloc& _Al, _CRT_GUARDOVERFLOW const size_type _Count) {
+#ifdef __cpp_lib_is_constant_evaluated // TRANSITION, GH-1532

```

Comment:
I am super sorry, I think you took that one from me. 

Above you guard destruction with `__cpp_lib_constexpr_dynamic_alloc`. I *think* this would be the right choice here too, as otherwise we do not need that code at all.

Please wait for a maintainers verdict before you go along and change all of those

Reply:
I think this change can also be done to both `allocate` overloads, `deallocate`, and `construct` below. Waiting for a maintainer's decision 🐱.

Reply:
If I'm understanding correctly, you are proposing to change the `#ifdef __cpp_lib_is_constant_evaluated` to `#ifdef __cpp_lib_constexpr_dynamic_alloc`?  I think we are in an interesting position where they function the same way here, so I don't know if it matters hugely.  

< (This is just me trying to reason through this, feel free to skip to the last paragraph if you already agree) >
The new behavior of potentially going through `_Al.(de)allocate` in `allocate` and `deallocate` only really needs to be applied if the functions are constant evaluated.  Because WG21-P0784 is the paper making the functions `constexpr`, if `__cpp_lib_is_constant_evaluated()` is defined and `_STD is_constant_evaluated() == true`, then we know `__cpp_lib_constexpr_dynamic_alloc` is defined as well.  If the function is not constant evaluated (whether that is because WG21-P0784 hasn't been implemented or just because it isn't called at compile-time), the old behavior should remain.  

This same logic applies to the change in `construct`.  We could just toggle `construct_at` vs. `::new` on `__cpp_lib_constexpr_dynamic_alloc` as we do in `destroy`, but here we eliminate a function call if the call is not made in `constexpr` land (the body of `construct_at` is the same as the pre-existing `::new` call).  Note that `destroy_at` actually has more logic than just calling a destructor (which is the pre-existing body of the function), which is why we just toggle based on `__cpp_lib_constexpr_dynamic_alloc` there.  

If we change to using `__cpp_lib_constexpr_dynamic_alloc` instead of `__cpp_lib_is_constant_evaluated`, we know that `_STD is_constant_evaluated()` is available, so there won't be any behavioral change there.  

In other words, I believe we would get the same behavior either way, so I think this question is reduced to a matter of convention.  I'm not aware of a strict convention here, but looking through the code base, I see primarily uses of `__cpp_lib_is_constant_evaluated` and `__cpp_lib_concepts`, and not many other feature test macros used.  This indicates that maybe we should stick with `__cpp_lib_is_constant_evaluated`, though I yield to another maintainer who feels more strongly.

Reply:
So my reasoning is slightly different and given that all supported compilers either support both or nothing it is kind of moot.

The point is that `__cpp_lib_constexpr_dynamic_alloc` is a stronger constraint than `__cpp_lib_is_constant_evaluated` in the sense that all compilers first implemented `__cpp_lib_is_constant_evaluated` and then `__cpp_lib_constexpr_dynamic_alloc`

So right now we are in the situation that a compiler that supports only `__cpp_lib_is_constant_evaluated` (MSVC right now) has to parse/validate those code snippets, although he can never use them. However, if we switch it around, only compilers that actually support `__cpp_lib_constexpr_dynamic_alloc` will see the code that is needed for it.

Does that make sense?

Reply:
Yeah, that makes sense.  I don't feel strongly either way - I am fine changing to `__cpp_lib_constexpr_dynamic_alloc` and whoever the second reviewer is can chime in if they strongly disagree :)

Reply:
I will change it to `__cpp_lib_constexpr_dynamic_alloc`  just from a consistency point of view in all affected member functions of `_Default_allocator_traits`. Hopefully, this will be the last change in this PR 😂.